pykafka/producer.py

```diff
@@ -540,7 +575,40 @@ def flush(self, linger_ms, release_pending=False):
         """
         self._wait_for_flush_ready(linger_ms)
         with self.lock:
-            batch = [self.queue.pop() for _ in range(len(self.queue))]
+            batch = []
+            batch_size_in_bytes = 0
+            while len(self.queue) > 0:
+                # TODO: Should we optimistically pop and renenqueue vs. peak
+                peaked_message = self.queue[-1]
+
+                if peaked_message and peaked_message.value is not None:
+                    if len(peaked_message) > max_request_size:
+                        exc = MessageSizeTooLarge(
+                            "Message size larger then max_request_size: {}".format(max_request_size)
+                        )
+                        log.warning(exc)
+                        # bind the MessageSizeTooLarge error the delivery
+                        # report and remove it from the producer queue
+                        message = self.queue.pop()
+                        peaked_message.delivery_report_q.put((message, exc))
+                        # remove from pending message count
+                        self.increment_messages_pending(-1)
+                        continue
+
+                    # test if adding the message would go over the
+                    # max_request_size. if it would, break out of loop
+                    elif batch_size_in_bytes + len(peaked_message) > max_request_size:
+                        log.debug("max_request_size reached. producing batch")
+                        # if we did not fully empty the queue. reset the
+                        # flush_ready so we send another batch immediately
+                        self.flush_ready.set()
+                        break
+
+                message = self.queue.pop()
+                if message.value is not None:
```

Comment:
I'm not sure this condition is necessary. `Message.__len__` takes the same condition into account, so I think it should be fine to remove this.